### PR TITLE
ENH: Use MaskedEngine for numeric pyarrow dtypes

### PR DESCRIPTION
**Pull request by phofl**
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

```
idx = pd.Index(list(range(1_000_000)), dtype="int64[pyarrow]")
idxer = list(range(1, 500_000, 10))

pr 
%timeit idx.get_indexer(idxer)
4.85 ms ± 24.4 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

main
%timeit idx.get_indexer(idxer)
16.9 ms ± 116 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

```

cc @mroeschke Is this roughly what you had in mind?

### CODE REVIEW

**Comment by mroeschke** [comment]
Nice! Yeah this is what I had in mind.

**Comment by phofl** [comment]
Added a whatsnew.

We should probably figure out a better way of converting to numpy when ``na_value`` is specified and compatible with the associated NumPy dtype

**Comment by phofl** [comment]
@mroeschke do we want to get this in before the rc?

File: pandas/_libs/index.pyx [hunk 1]
```diff
@@ -1141,12 +1141,24 @@ cdef class ExtensionEngine(SharedEngine):
...
-        super().__init__(values._data)
-        self.mask = values._mask
+        super().__init__(self._get_data(values))
+        self.mask = self._get_mask(values)
+
+    def _get_data(self, object values) -> np.ndarray:
+        if hasattr(values, "_mask"):
+            return values._data
+        # We are an ArrowExtensionArray
+        return values.to_numpy(na_value=1, dtype=values.dtype.numpy_dtype)
```
Comment (mroeschke) [inline pandas/_libs/index.pyx:1151]: Curious why do we specify `na_value=1`?

Comment (phofl) [reply pandas/_libs/index.pyx:1151]: We have to use something that fits into all numeric arrays, otherwise we geht NA and end up with object

Comment (mroeschke) [reply pandas/_libs/index.pyx:1151]: Got it, could you add a comment to that effect here? With that and a rebase this LGTM

Comment (mroeschke) [reply pandas/_libs/index.pyx:1151]: Also a TODO to remove this once we have a separate indexing engine for pyarrow

Comment (phofl) [reply pandas/_libs/index.pyx:1153]: Done

**Pull request merged**